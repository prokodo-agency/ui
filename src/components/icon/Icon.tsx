import {
  lazy,
  Suspense,
  useMemo,
  type FC,
  type ComponentType,
} from "react"
import { create } from "@/helpers/bem"
import styles from "./Icon.module.scss"
import type { IconProps, IconSize } from "./Icon.model"
import { ICONS } from "./iconsMap"

const bem = create(styles, "Icon")

const FallbackIcon: FC = () => (
  <svg viewBox="0 0 24 24" height="24" width="24">
    <circle cx="12" cy="12" r="10" stroke="black" strokeWidth="2" />
    <line stroke="black" strokeWidth="2" x1="6" y1="6" x2="18" y2="18" />
  </svg>
)

export const getIconSize = (size?: IconSize): number => {
  switch (size) {
    case "xs":
      return 15
    case "sm":
      return 20
    case "md":
      return 30
    case "lg":
      return 40
    case "xl":
      return 50
    default:
      return size ?? 16
  }
}

// Converts PascalCase to snake_case used in filenames
const formatIconName = (name: string): string =>
  name
    .replace(/([a-z])([A-Z])/g, "$1_$2")
    .replace(/Icon$/, "")
    .toLowerCase()

const getIconLoader = (name: string) => {
  const snakeName = formatIconName(name)
  const loader = ICONS[`${snakeName}_icon`]
  return loader ? loader() : Promise.resolve({ default: FallbackIcon })
}

export const Icon: FC<IconProps> = ({
  name,
  size,
  color,
  className,
  ...props
}) => {
  const LazyIcon = useMemo(
    () =>
      lazy(async () => {
        const Component = await getIconLoader(name ?? "")
        return { default: Component as ComponentType<any> }
      }),
    [name]
  )

  return (
    <Suspense fallback={<FallbackIcon />}>
      <LazyIcon
        aria-hidden="true"
        role="presentation"
        size={getIconSize(size)}
        color={color}
        name={name}
        {...props}
        className={bem(
          undefined,
          typeof color === "string" ? { [color]: true } : undefined,
          className
        )}
      />
    </Suspense>
  )
}
