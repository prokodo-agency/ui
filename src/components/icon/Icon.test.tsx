import { expect } from "@jest/globals"

import { render, screen } from "@/tests"

import { Icon } from "./Icon"

import type { FC, SVGProps } from "react"

jest.mock("@/assets/icons/tire.svg", () => {
  const svg = "<svg><text>tire icon</text></svg>"
  const ReactComponent: FC<SVGProps<SVGSVGElement>> = props => (
    <svg {...props}>
      <text>tire icon</text>
    </svg>
  )

  return {
    default: svg,
    ReactComponent,
  }
})

describe("The common icon component", () => {
  it("should render a svg icon", () => {
    render(<Icon name="AbacusIcon" />)
    const svgIcon = screen.getByRole("img") // Assuming the svg has the role "img"
    expect(svgIcon).toMatchSnapshot()
  })

  it("should render with default accessibility attributes", () => {
    render(<Icon name="AbsoluteIcon" />)
    const svgIcon = screen.getByRole("img", { hidden: true }) // For hidden elements
    expect(svgIcon).toHaveAttribute("role", "presentation")
    expect(svgIcon).toHaveAttribute("aria-hidden", "true")
  })

  it("should allow to overwrite accessibility attributes", () => {
    render(<Icon aria-hidden="false" name="AccessIcon" role="alert" />)
    const svgIcon = screen.getByRole("alert") // Now querying by the role set to "alert"
    expect(svgIcon).toHaveAttribute("role", "alert")
    expect(svgIcon).toHaveAttribute("aria-hidden", "false")
  })
})
