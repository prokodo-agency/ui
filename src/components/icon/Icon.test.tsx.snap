// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The common icon component should render a svg icon 1`] = `
<svg
  aria-hidden="true"
  fill="none"
  height="16"
  role="presentation"
  stroke="currentColor"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="2"
  viewBox="0 0 24 24"
  width="16"
  xmlns="http://www.w3.org/2000/svg"
>
  <line
    x1="18"
    x2="6"
    y1="6"
    y2="18"
  />
  <line
    x1="6"
    x2="18"
    y1="6"
    y2="18"
  />
</svg>
`;
